packages/mangrove-arbitrage/src/MgvArbitrage.sol

```diff
@@ -1,55 +1,82 @@
+  IERC20 takerGivesToken;
+  IERC20 takerWantsToken;
+  uint tickSpacing;
+  IUniswapV3Pool pool;
+  uint160 minimumGain; // if zero then ignored
 }
 
-contract MgvArbitrage is AccessControlled {
+contract MgvArbitrage is AccessControlled, IUniswapV3SwapCallback {
+  /// copied from v3-core repo to interact with low level swap function of uniwap
+  uint160 internal constant MIN_SQRT_RATIO = 4295128739;
+  uint160 internal constant MAX_SQRT_RATIO = 1461446703485210103287273052203988822378723970342;
+
   IMangrove public mgv;
-  address public arbitrager;
+
+  bool isArbitraging;
+
+  mapping(address => bool) public pools;
 
   /// @param _mgv The Mangrove instance to be arbitraged
   /// @param admin The admin of the contract. The only address allowed to withdraw funds.
-  constructor(IMangrove _mgv, address admin, address _arbitrager) AccessControlled(admin) {
+  constructor(IMangrove _mgv, address admin) AccessControlled(admin) {
     mgv = _mgv;
-    arbitrager = _arbitrager;
   }
 
-  /// @notice This modifier verifies that the `msg.sender` is the arbitrager
-  modifier onlyArbitrager() {
-    require(msg.sender == arbitrager, "Only arbitrager can call this function");
-    _;
-  }
+  receive() external payable virtual {}
 
-  /// @notice The router used to do the swap on Uniswap
-  ISwapRouter public router = ISwapRouter(0xE592427A0AEce92De3Edee1F18E0157C05861564);
+  /// copied from there: https://ethereum.stackexchange.com/a/136961
+  function getPrice(IUniswapV3Pool pool, uint token0Decimals) internal view returns (uint price) {
+    (uint160 sqrtPriceX96,,,,,,) = pool.slot0();
+    uint numerator1 = uint(sqrtPriceX96) * uint(sqrtPriceX96);
+    uint numerator2 = 10 ** token0Decimals;
+    return FullMath.mulDiv(numerator1, numerator2, 1 << 192);
+  }
+
+  /// @notice Estimate the amount of a token you will receive from a Uniswap V3 pool for a given input amount.
+  /// @param pool The Uniswap V3 pool you want to estimate the trade in.
+  /// @param takerGivesToken The token you provide as input.
+  /// @param balance The amount of the input token you want to trade.
+  /// @return amount The estimated amount of the other token you will receive.
+  function estimateHowMuchUniswapV3CanGive(IUniswapV3Pool pool, IERC20 takerGivesToken, uint balance)
+    internal
+    view
+    returns (uint amount)
+  {
+    IERC20 token0 = IERC20(pool.token0());
 
-  receive() external payable virtual {}
+    uint token0Decimals = token0.decimals();
+    uint price = getPrice(pool, token0Decimals);
 
-  /// @notice This enables the admin to change the arbitrager.
-  /// @param newArbitrager The new arbitrager
-  function setArbitrager(address newArbitrager) external onlyAdmin {
-    arbitrager = newArbitrager;
+    if (token0 != takerGivesToken) {
+      // price is in takerWantsToken
+      amount = 10 ** token0Decimals * balance / price;
+    } else {
+      // price is in takerGivesToken
+      amount = (balance * price) / 10 ** token0Decimals;
+    }
```

Comment:
Assuming infinite liquidity on uniswap. Do we know for what volume is the price true ? Can't we estimate the precise output of a swap instead ?

Reply:
No but we don't really care,
This is just to give us a signal if we should either SELL all token A or BUY exact amount of token B. As we are arbitraging liquid pools it should not be an issue. 

The worst case scenario would be that we miss one arbitrage opportunity.